src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "asprof.h"
+#include "assert.h"
+#include "codeCache.h"
+#include "mallocTracer.h"
+#include "os.h"
+#include "profiler.h"
+#include <dlfcn.h>
+#include <string.h>
+
+#define ADDRESS_OF(sym) ({               \
+    void* addr = dlsym(RTLD_NEXT, #sym); \
+    addr != NULL ? (sym##_t)addr : sym;  \
+})
+
+typedef void* (*malloc_t)(size_t);
+static malloc_t _orig_malloc = NULL;
+
+typedef void* (*calloc_t)(size_t, size_t);
+static calloc_t _orig_calloc = NULL;
+
+typedef void* (*realloc_t)(void*, size_t);
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_calloc(1, size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* malloc(size_t size) {
+    if (unlikely(!_orig_malloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return malloc_hook(size);
+    }
+    return _orig_malloc(size);
+}
+#endif // __linux__
+
+static void* calloc_hook(size_t num, size_t size) {
+    void* ret = _orig_calloc(num, size);
+    if (likely(ret && num && size)) {
+        MallocTracer::recordMalloc(ret, num * size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* calloc(size_t num, size_t size) {
+    if (unlikely(!_orig_calloc)) {
+        // In some libc versions, dlsym may call calloc on error case in the ADDRESS_OF macro.
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return calloc_hook(num, size);
+    }
+    return _orig_calloc(num, size);
+}
+#endif // __linux__
+
+static void* realloc_hook(void* addr, size_t size) {
+    void* ret = _orig_realloc(addr, size);
+    if (likely(ret && addr)) {
+        MallocTracer::recordFree(addr, OS::nanotime());
+    }
+
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* realloc(void* addr, size_t size) {
+    if (unlikely(!_orig_realloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return realloc_hook(addr, size);
+    }
+    return _orig_realloc(addr, size);
+}
+#endif // __linux__
+
+static void free_hook(void* addr) {
+    _orig_free(addr);
+    MallocTracer::recordFree(addr, OS::nanotime());
```

Comment:
`free` can be legitimately called for NULL. No need to record an event for this, I think.